agent/wsclient/client_test.go

```diff
@@ -0,0 +1,102 @@
+// Copyright 2014-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package wsclient
+
+import (
+	"os"
+	"testing"
+	"time"
+
+	"github.com/aws/amazon-ecs-agent/agent/acs/model/ecsacs"
+	"github.com/aws/amazon-ecs-agent/agent/wsclient/mock/utils"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+)
+
+// Concurrent writes will cause a panic in the websocket library
+// See https://godoc.org/github.com/gorilla/websocket#hdr-Concurrency
+func TestConcurrentWritesDontPanic(t *testing.T) {
+	mockServer, _, requests, _, _ := mockwsutils.StartMockServer(t, make(chan bool))
+	defer mockServer.Close()
+
+	req := ecsacs.AckRequest{Cluster: aws.String("test"), ContainerInstance: aws.String("test"), MessageId: aws.String("test")}
+
+	cs := getClientServer(mockServer.URL)
+	cs.Connect()
+
+	executeTenRequests := func() {
+		for i := 0; i < 10; i++ {
+			cs.MakeRequest(&req)
+		}
+	}
+
+	go executeTenRequests()
+	executeTenRequests()
+
+	timeout := make(chan bool)
+	go func() {
+		time.Sleep(5 * time.Second)
+	}()
+
+	//Ensure all 20 requests succeed
+	for i := 0; i < 20; i++ {
+		select {
+		case <-requests:
+			continue
+		case <-timeout:
+			t.Fatal("Wanted 20 requests, only received:", i)
+		}
+	}
+}
+
+func TestProxyVariableCustomValue(t *testing.T) {
+	mockServer, _, _, _, _ := mockwsutils.StartMockServer(t, make(chan bool))
+	defer mockServer.Close()
+
+	testString := "Custom no proxy string"
+	os.Setenv("DOCKER_HOST", "/var/run/docker.sock")
+	os.Setenv("NO_PROXY", testString)
+	getClientServer(mockServer.URL).Connect()
+
+	if os.Getenv("NO_PROXY") != testString {
```

Comment:
(multiple occurrences) Can you use the `github.com/stretchr/testify/assert` library?

```go
assert.Equal(t, testString, os.Gentenv("NO_PROXY"), "NO_PROXY should be overwritten when non-default value is supplied")
```